web_src/js/webcomponents/absolute-date.test.ts

```diff
@@ -7,9 +7,15 @@ test('toAbsoluteLocaleDate', () => {
     day: 'numeric',
   })).toEqual('March 15, 2024');
 
-  expect(toAbsoluteLocaleDate('2024-03-15', 'de-DE', {
+  expect(toAbsoluteLocaleDate('2024-03-15T01:02:03', 'de-DE', {
     year: 'numeric',
     month: 'long',
     day: 'numeric',
   })).toEqual('15. März 2024');
```

Comment:
What is the output of the function for input `2024-03-15` now?

There are cases where only date without time is passed, for example for milestones. I think this change may break those cases.

Reply:
![image](https://github.com/user-attachments/assets/af3e657a-82d1-40d1-ac7b-0ed5dfa0b45f)

Reply:
![image](https://github.com/user-attachments/assets/3d7c031c-3bb7-45b0-bd38-c5c612c0ac07)

![image](https://github.com/user-attachments/assets/d2252495-3d8e-4edc-a208-41098c25ffed)

![image](https://github.com/user-attachments/assets/6b72fc55-47f6-4317-9567-b55a96e70d7b)

Reply:
The problem is:

```
> new Date("2024-03-15")
Date Fri Mar 15 2024 01:00:00 GMT+0100 (Central European Standard Time)
```

This date object has a time component which is wrong because dates are timeless. Is it guaranteed that the output of the following is always the same, regardless of time and timezone?

```
> new Date("2024-03-15").toLocaleString("de-DE", {year: "numeric", month: "long", day: "numeric"})
15. März 2024
```

Reply:
It is not related.

Using a backend date string is wrong, for long time. That string is in server's timezone, is a string, doesn't have the timezone field, and frontend don't know, it would definitely go wrong when the browser's timezone is different from the server's.

To make the "deadline" right, it needs to completely refactor the backend first.

ps: the frontend use the string "date" part is also wrong, but I didn't change that logic.

Reply:
And one more thing, the `"de-DE"` you mentioned is "locale", not "timezone". They are not related.

Reply:
This PR only fixed 2 bugs in old code:

1. `this.getAttribute('date').substring(0, 10);` causes corrupted date string and then causes errors
2. `lang ?? []` should be `lang || []`, otherwise empty lang causes errors.

Other logics are kept as-is. Using `Date` here should be the same as `Temporal.PlainDate`, unless you could show a real bad case.

Reply:
> Using a backend date string is wrong, for long time. That string is in server's timezone

Things like milestone due dates are stored as `yyyy-mm-dd` in the backend and this is correct and does not need to change. It's just a date, no time or timezone.

Using `Date` on these values is not correct because `Date` is meant for timestamps with time and timezone. `Temporal.PlainDate` is specifically meant for such dates.

Reply:
> unless you could show a real bad case.

Reply:
> Other logics are kept as-is. Using Date here should be the same as Temporal.PlainDate, unless you could show a real bad case.

I wasn't able to produce a case where it outputs wrong date so far, but it might be possible with certain local time or timezone values.

Reply:
But the server DOES part it, especially when a user sets the deadline. The server doesn't know the user's timezone, it is already wrong in this step.

To completely fix the problem, we should make the "deadline" supports  time and timezone.

The first step is:  Replace DateTime with DateUtils #32383

Reply:
If more precise deadline is desired (and I think it might make sense), then yes, it should be handled as a as timestamp with time and timezone components in both backend and frontend.

So far I was assuming deadline was supposed to be a plain date.

Reply:
> If more precise deadline is desired (and I think it might make sense), then yes, it should be handled as a as timestamp with time and timezone components in both backend and frontend and the UI should also allow picking a time in addition to the current date-only input.

Let's do it, the first step is  #32383, and I think I can finish the whole refactoring in one or two days.

Reply:
> Why not? new Date("2024-03-15") is in current timezone, so it must be 15. März 2024

> We do not fill time or timezone into Date, it only uses the day in current timezone. So would there really be any problem?

As per https://tc39.es/ecma262/multipage/numbers-and-dates.html#sec-date-time-string-format, the "date-only form" seems to at least be specced, so I guess it will be okay in all major JS engines. Historically, the `Date` constructor was never really specced so still has varying behaviour across JS engines. I would personally only use the Temporal API for new projects instead to be safe.

Reply:
There is also this note in the spec:

> When the UTC offset representation is absent, date-only forms are interpreted as a UTC time and date-time forms are interpreted as a local time.

So if I understand correct, it should interpret a plain date with absent offset as 00:00 UTC, which would result in the wrong date being rendered if the user is in a timezone with a negative offset to UTC.

Reply:
Confirmed on Linux:

```
$ timedatectl set-timezone US/Eastern
$ node -p 'new Date("2024-03-15").toLocaleString("de-DE", {year: "numeric", month: "long", day: "numeric"})'
14. März 2024
```

So yes, please revert to `Temporal.PlainDate`.

Reply:
Then we only need to do `new Date(date + 'T00:00:00')`, then the result is right.

Reply:
Even easier reproduction, this should be possible to do in unit tests:

```
> process.env.TZ = "US/Eastern"; new Date("2024-03-15").toLocaleString("de-DE", {year: "numeric", month: "long", day: "numeric"})
'14. März 2024'
> process.env.TZ = "Europe/Berlin"; new Date("2024-03-15").toLocaleString("de-DE", {year: "numeric", month: "long", day: "numeric"})
'15. März 2024'
```

Reply:
> Then we only need to do `new Date(date + 'T00:00:00')`, then the result is right.

Yes, but why abuse an API that was not made for this use case? `Temporal.PlainDate` was made for this.

Reply:
-> Fix toAbsoluteLocaleDate and add more tests #32387 

Added more tests.

----

> `Temporal.PlainDate` was made for this.

It just introduces more unnecessary polyfills, make the code unnecessary larger, and according to MDN: **It is not yet ready for production use!** (just quoted the sentence).

Reply:
> It is not yet ready for production use

Right, it's still in stage 3, so not finalized and will probably take another year or two until we see it being implemented in browsers.